frontend/src/pages/PipelineDetails.tsx

```diff
@@ -235,24 +239,49 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
     if (origin) {
       const msgRunOrRecurringRun = origin.isRecurring ? 'recurring run' : 'run';
       try {
+        // TODO(jlyaoyuli): change to v2 API after v1 is deprecated
         if (origin.isRecurring) {
-          origin.recurringRun = await Apis.jobServiceApi.getJob(origin.recurringRunId!);
+          origin.v1RecurringRun = await Apis.jobServiceApi.getJob(origin.recurringRunId!);
+          origin.v2RecurringRun = await Apis.recurringRunServiceApi.getRecurringRun(
+            origin.recurringRunId!,
+          );
         } else {
-          origin.run = await Apis.runServiceApi.getRun(origin.runId!);
+          origin.v1Run = await Apis.runServiceApi.getRun(origin.runId!);
+          origin.v2Run = await Apis.runServiceApiV2.getRun(origin.runId!);
+        }
+
+        // v2 run or recurring run with pipeline_version_id
+        const pipelineVersionIdFromOrigin = origin.isRecurring
+          ? origin.v2RecurringRun?.pipeline_version_id
+          : origin.v2Run?.pipeline_version_id;
+
+        let templateStrFromOrigin = '';
+        if (pipelineVersionIdFromOrigin) {
+          const response = await Apis.pipelineServiceApi.getPipelineVersionTemplate(
+            pipelineVersionIdFromOrigin,
+          );
+          templateStrFromOrigin = response.template || '';
+        }
+
+        // v2 run or recurring run with pipeline_spec
+        let pipelineManifest = '';
+        if (origin.v2Run?.pipeline_spec) {
+          pipelineManifest = JsYaml.safeDump(origin.v2Run.pipeline_spec);
+        }
+        if (origin.v2RecurringRun?.pipeline_spec) {
+          pipelineManifest = JsYaml.safeDump(origin.v2RecurringRun.pipeline_spec);
         }
```

Comment:
This piece of logic is getting complicated. I suggest to abstract it to a function. Ultimately what it tries to do is simple: Get pipeline template from run or recurring run.